xbmc/playlists/PlayListXSPF.cpp

```diff
@@ -0,0 +1,188 @@
+      <info>https://kexp.org/</info>
+      <image>https://www.kexp.org/static/assets/img/logo-white.svg</image>
+		</track>
+    <track><location>Relative/Path/To.mp3</location></track>
+    <track><location>file:///music/song_1.ogg</location></track>
+    <track><location>file:///music/song_2.flac</location></track>
+    <track><location>file:///music/song_3.mp3</location></track>
+	</trackList>
+</playlist>
+*/
+
+#define FILE_PROTOCOL_WITH_ROOT "file:///"
+#define LOCATION_TAGNAME "location"
+#define PLAYLIST_TAGNAME "playlist"
+#define TITLE_TAGNAME "title"
+#define TRACK_TAGNAME "track"
+#define TRACKLIST_TAGNAME "trackList"
+
+#define missingElementMessage(E) "no " E " element"
+
+
+CPlayListXSPF::CPlayListXSPF(void) = default;
+
+CPlayListXSPF::~CPlayListXSPF(void) = default;
+
+
+bool CPlayListXSPF::Load(const std::string& strFileName)
+{
+  CXBMCTinyXML xmlDoc;
+  const char * errorStr = "";
+  int errorRow = 0;
+  int errorCol = 0;
+
+  const char * label = "";
+  const char * location = "";
+  int offset = 0;
+  std::string localpath;
+  CURL uri;
+
+  TiXmlElement* pPlaylist = nullptr;
+  TiXmlElement* pTracklist = nullptr;
+  TiXmlElement* pPlaylistTitle = nullptr;
+  TiXmlElement* pCurTrack = nullptr;
+
+  CLog::Log(LOGINFO, "ZZ Playlist=%s", strFileName);
+
+  if (!xmlDoc.LoadFile(strFileName))
+  {
+    goto error;
+  }
+
+  pPlaylist = xmlDoc.FirstChildElement(PLAYLIST_TAGNAME);
+  if (!pPlaylist)
+  {
+    errorStr = missingElementMessage(PLAYLIST_TAGNAME);
+    goto error;
+  }
+
+  pTracklist = pPlaylist->FirstChildElement(TRACKLIST_TAGNAME);
+  if (!pTracklist)
+  {
+    errorRow = pPlaylist->Row();
+    errorCol = pPlaylist->Column();
+    errorStr = missingElementMessage(TRACKLIST_TAGNAME);
+    goto error;
+  }
+
+  Clear();
+  URIUtils::GetParentPath(strFileName, m_strBasePath);
+
+  pPlaylistTitle = pPlaylist->FirstChildElement(TITLE_TAGNAME);
+  if (pPlaylistTitle)
+  {
+    m_strPlayListName = pPlaylistTitle->GetText();
+  }
+
+  pCurTrack = pTracklist->FirstChildElement(TRACK_TAGNAME);
+
+  while (pCurTrack)
+  {
+    label = "";
+    location = "";
+    localpath.clear();
+    uri.Reset();
+
+    TiXmlElement* pCurLocation = pCurTrack->FirstChildElement(LOCATION_TAGNAME);
+    if (pCurLocation)
+    {
+
+      TiXmlElement* pCurTitle = pCurTrack->FirstChildElement(TITLE_TAGNAME);
+      if (pCurTitle)
+      {
+        label = pCurTitle->GetText();
+      }
+
+      CFileItemPtr newItem(new CFileItem(label));
+
+      location = pCurLocation->GetText();
+      CLog::Log(LOGINFO, "Interpreting location=%s", location);
+
+      uri.Parse(location);
+
+      if (StringUtils::StartsWith(location, FILE_PROTOCOL_WITH_ROOT))
+      {
+#ifdef TARGET_WINDOWS
+        // omit leading / from "file:///C:/Users/..."
+	      offset = sizeof(FILE_PROTOCOL_WITH_ROOT) - 1;
+#else
+        // include leading / from "file:///home/..."
+	      offset = sizeof(FILE_PROTOCOL_WITH_ROOT) - 2;
+#endif
```

Comment:
this looks hacky
indentation is wrong

Reply:
Dang, that indentation slipped by. As for the pattern what's going on is the "file://" [URI scheme](https://en.wikipedia.org/wiki/File_URI_scheme) has no host and goes straight into an absolute path. On Windows the path needs to start at the drive and on Linux the path needs to start with the leading "/". Since the "/" is already in the string I opted to grab it. What do you think of this, instead:

```c++
#ifdef TARGET_WINDOWS
	localpath = "";
#else
        localpath = "/";
#endif
        localpath += CURL::Decode(location + sizeof(FILE_PROTOCOL_WITH_ROOT) - 1);
```

Reply:
looks better
please avoid pointer arithmetic with and also the sizeof, better use length and substr method from std::string

Reply:
Ah, yes, that is cleaner and more obvious, thanks :)

Reply:
I removed the `sizeof` math off a pointer in lieu of substr with a number and a comment. (I tried to use URL.cpp parsing as a style-guide).